include/wil/async.h

```diff
@@ -0,0 +1,139 @@
+//    TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+//    PARTICULAR PURPOSE AND NONINFRINGEMENT.
+//
+//*********************************************************
+#ifndef __WIL_ASYNC_INCLUDED
+#define __WIL_ASYNC_INCLUDED
+
+#include <version>
+#ifndef __cpp_lib_coroutine
+#error This header requires C++20 coroutines.
+#endif
+
+#include <coroutine>
+#include <Windows.h>
+#include "resource.h"
+#include "result.h"
+
+namespace wil
+{
+    namespace details
+    {
+        struct operation_info
+        {
+            OVERLAPPED overlapped = { };
+
+            bool completed = false;
+            HRESULT error = S_OK;
+
+            std::coroutine_handle<> coro;
+        };
+
+        template<typename err_policy>
+        struct async_file_operation
+        {
+            unique_process_heap_ptr<operation_info> info = nullptr;
+            HRESULT setup_error = S_OK;
+
+            bool await_ready() const noexcept
+            {
+                return FAILED(setup_error) || info->completed;
+            }
+
+            typename err_policy::result await_resume() const
+            {
+                if (FAILED(setup_error))
+                {
+                    return err_policy::HResult(setup_error);
+                }
+                else
+                {
+                    return err_policy::HResult(info->error);
+                }
+            }
+
+            void await_suspend(std::coroutine_handle<> handle) noexcept
+            {
+                info->coro = handle;
+            }
+        };
+
+        void WINAPI OverlappedCallback(DWORD error, DWORD count, LPOVERLAPPED overlapped)
+        {
+            const auto info = reinterpret_cast<details::operation_info*>(overlapped);
+            info->completed = true;
+            info->error = HRESULT_FROM_WIN32(error);
+
+            if (info->overlapped.hEvent)
+            {
+                *static_cast<DWORD*>(info->overlapped.hEvent) = count;
+            }
+
+            if (info->coro)
+            {
+                info->coro();
+            }
+        }
+    }
+
+    template<typename err_policy = err_exception_policy>
+    details::async_file_operation<err_policy> ReadFileAsync(HANDLE handle, LPCVOID buffer, DWORD bytes_to_read, DWORD* bytes_read, UINT64 offset)
+    {
+        details::async_file_operation<err_policy> op;
+        op.info.reset(new (::HeapAlloc(::GetProcessHeap(), 0, sizeof(details::operation_info))) details::operation_info);
+        if (op.info)
+        {
+            // this is not used by ReadFileEx so we'll use it.
+            op.info->overlapped.hEvent = bytes_read;
+
+            ULARGE_INTEGER offset_large;
+            offset_large.QuadPart = offset;
+            op.info->overlapped.Offset = offset_large.LowPart;
+            op.info->overlapped.OffsetHigh = offset_large.HighPart;
+
+            if (::ReadFileEx(handle, buffer, bytes_to_read, &op.info->overlapped, details::OverlappedCallback))

```

Comment:
While I'm all for the idea of this PR, there's a problem here in that it only works if, like in your sample, the caller making the async request actually blocks in an alertable state. The callback for the completion ONLY happens ON the thread that made the original call to ReadFileEx AND only while in a alertable state. That's a restriction with a lot of surprise attached to it. If the call stack doing the co_await never blocks in an alertable state (Surprisingly common as not all synch objects support this and the default is not alertable for most that do)

I'm not sure if documentation alone is enough to resolve that and if there could be subtle issues that complicate things. That is, in local dev loop testing everything completes synchronously, but in the field it doesn't subtle behavior differences could cause difficult to track bugs. That said, though it wouldn't actually be worse than what you can get into using the APIs raw...

Reply:
I could rename the functions to `wil::ReadFileEx` and `wil::WriteFileEx`, that would make it clearer that the alertable/threading behavior matches the native APIs with the same name. I am not very familiar with IOCPs and there are even more ways to do async IO (like `RegisterWaitForSingleObject`).

As a note to myself, this is missing some `GetOverlappedResult` calls.

Reply:
Rename would help clarify this particular case. Though, I'm not sure that wrapping ReadFileEx/WriteFileEx is really the way to go for Async IO in WIL. As you say there are a lot of ways to do ASYNC IO in Win32, each having advantages. IOCompletion ports are another option. There is also the Thread Pool IO support APIs that actually use IO Completion ports under the hood but hide that in a a way that likely fits the coroutine co_await style paradigm better. (see: [CreateThreadPoolIo](https://docs.microsoft.com/en-us/windows/win32/api/threadpoolapiset/nf-threadpoolapiset-createthreadpoolio) and related APIs)